### PR TITLE
updated charfluctuations analysis

### PR DESCRIPTION
**Pull request by mlund**
Before sending a pull request, please ensure that:

- all tests pass, i.e. run `make test`
- all compiler warnings have been fixed
- source files use soft-indendation of four
- no temporary files (.o, .pqr, state etc) are included in the commit

### CODE REVIEW

**Comment by mlund** [comment]
Hi @gitesei, looks good. The code can be simplified by allowing for only _one_ molecule in the input, i.e. `molecules` -> `molecule`. If the user needs to analyse several molecules, it's merely to add another analysis instance in the input file.

File: src/analysis.h [hunk 1]
```diff
@@ -736,58 +736,94 @@ namespace Faunus {
...
+                bool verbose = false;
+                size_t cnt;
 
                 void _sample() override {
-                    cnt = 0;
-                    for (auto p=spc.p.begin()+ndx0; p!=(spc.p.begin()+ndx1+1); ++p) {
-                        idcnt.at(cnt)[p->id]++;
-                        charge.at(cnt) += p->charge;
-                        ++cnt;
+                    for (auto &g : spc.groups) {
```
Comment (mlund) [inline src/analysis.h]: ```for (auto &g : spc.findMolecules(molid) {...}``` would simplify

File: src/analysis.h [hunk 2]
```diff
@@ -736,58 +736,94 @@ namespace Faunus {
...
+                        cnt = 0;
+                        if (g.id == molid) {
+                            for ( auto p = g.begin(); p < g.trueend() ; ++p) {
```
Comment (mlund) [inline src/analysis.h]: Should this go to `trueend()`?

File: src/analysis.h [hunk 3]
```diff
@@ -736,58 +736,94 @@ namespace Faunus {
...
+                    molname = j.at("molecule").get<decltype(molname)>(); // molecule name
+                    auto it = findName(Faunus::molecules<Tpvec>, molname);
+                    if (it == Faunus::molecules<Tpvec>.end())
+                        throw std::runtime_error("unknown species '" + molname + "'");
+                    molid = it->id(); // molecule ID
+                    idcnt.resize( it->atoms.size() );
+                    charge.resize( it->atoms.size() );
+                }
+
+                virtual ~ChargeFluctuations() {
```
Comment (mlund) [inline src/analysis.h]: cleaner to use the `FormatPQR` from io.h. I think it takes a vector of particles.

File: docs/_docs/analysis.md [hunk 4]
```diff
@@ -228,14 +228,17 @@ The input keywords are:
 
 ## Charge Fluctuations
 
-Calculates the average charge, the standard deviation of the charge, and the most 
-probable species (atom name) of atomic indexes undergoing charge fluctuations.
+For a given molecule, this calculates the average charge, the standard deviation of the charge, 
+and the most probable species (atom name) averaging over all molecular groups.
+The structure of one of the molecular groups with average charges and most probable 
+atomic species is saved in PQR format.
```
Comment (mlund) [inline docs/_docs/analysis.md]: Use instead a `pqrfile` keyword to let the user decide the name and if a pqr file should be saved?

File: src/analysis.h (hunk 2 above)
Comment (gitesei) [reply src/analysis.h]: No, I agree. I got confused and here it should be auto &p : g. Later in the code, when I save the PQR file I should instead loop to the trueend() to make sure that the whole molecule is saved.

**Pull request merged**